File: app/controllers/involvements_controller.rb
class InvolvementsController < ApplicationController
  protect_from_forgery with: :exception

  def create
    attrs = involvement_params.merge(participant_type: "Soldier")
    inv = Involvement.find_or_initialize_by(
      participant_type: attrs[:participant_type],
      participant_id:   attrs[:participant_id],
      involvable_type:  attrs[:involvable_type],
      involvable_id:    attrs[:involvable_id]
    )
    inv.role = attrs[:role]
    inv.year = attrs[:year]
    inv.note = attrs[:note]

    if inv.save
      render json: involvement_json(inv), status: :created
    else
      render json: { errors: inv.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotUnique
    inv = Involvement.find_by!(
      participant_type: attrs[:participant_type],
      participant_id:   attrs[:participant_id],
      involvable_type:  attrs[:involvable_type],
      involvable_id:    attrs[:involvable_id]
    )
    inv.update(role: attrs[:role], year: attrs[:year], note: attrs[:note])
    render json: involvement_json(inv), status: :ok
  end

  def destroy
    inv = Involvement.find(params[:id])
    inv.destroy
    head :no_content
  end

  private

  def involvement_params
    params.require(:involvement).permit(
      :involvable_type, :involvable_id,
      :participant_id,
      :role, :year, :note
    )
  end

  def involvement_json(inv)
    {
      id: inv.id,
      participant_id:   inv.participant_id,
      participant_type: inv.participant_type,
      participant_label: inv.participant_label,
      role: inv.role,
      year: inv.year,
      note: inv.note
    }
  end
end

File: app/javascript/controllers/involvements_controller.js
// app/javascript/controllers/involvement_form_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["form", "list", "participantId", "role", "year", "note"]

  async add(event) {
    event.preventDefault()
    const token = document.querySelector('meta[name="csrf-token"]').content
    const body = {
      involvement: {
        participant_id: this.participantIdTarget.value,
        involvable_type: this.formTarget.dataset.involvableType,
        involvable_id: this.formTarget.dataset.involvableId,
        role: this.roleTarget.value,
        year: this.yearTarget.value,
        note: this.noteTarget.value
      }
    }
    const res = await fetch("/involvements.json", {
      method: "POST",
      headers: { "Content-Type": "application/json", "X-CSRF-Token": token, "Accept": "application/json" },
      body: JSON.stringify(body)
    })
    if (res.ok) {
      const inv = await res.json()
      this.appendRow(inv)
      this.formTarget.reset()
      this.participantIdTarget.value = ""
    } else {
      const err = await res.json().catch(() => ({}))
      alert((err.errors && err.errors.join(", ")) || "Could not add involvement.")
    }
  }

  async remove(event) {
    event.preventDefault()
    const id = event.params.id // from data-action param
    const token = document.querySelector('meta[name="csrf-token"]').content
    const res = await fetch(`/involvements/${id}.json`, {
      method: "DELETE",
      headers: { "X-CSRF-Token": token, "Accept": "application/json" }
    })
    if (res.status === 204) {
      this.element.querySelector(`[data-involvement-row="${id}"]`)?.remove()
    } else {
      alert("Could not remove involvement.")
    }
  }

  appendRow(inv) {
    const li = document.createElement("li")
    li.dataset.involvementRow = inv.id
    li.className = "mb-2"
    const bits = []
    bits.push(`<a href="/${inv.participant_type.toLowerCase()}s/${inv.participant_id}">${inv.participant_label}</a>`)
    if (inv.role) bits.push(` — <strong>${this.escape(inv.role)}</strong>`)
    if (inv.year) bits.push(` (${this.escape(inv.year)})`)
    if (inv.note) bits.push(` — <em>${this.escape(inv.note)}</em>`)
    bits.push(` <button data-action="involvement-form#remove" data-involvement-form-id-param="${inv.id}" class="btn btn-sm btn-outline-danger ms-2">Remove</button>`)
    li.innerHTML = bits.join("")
    this.listTarget.appendChild(li)
  }

  escape(s) { return String(s).replace(/[&<>"']/g, m => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[m])) }
}
